src/OpenFeature/EventExecutor.cs

```diff
@@ -0,0 +1,294 @@
+                this._activeSubscriptions.Remove(oldProvider);
+                var channel = oldProvider.Provider.GetEventChannel();
+                if (channel != null)
+                {
+                    channel.Writer.WriteAsync(new ShutdownSignal());
+                }
+            }
+        }
+
+        private bool IsProviderBound(FeatureProviderReference provider)
+        {
+            if (this._defaultProvider == provider)
+            {
+                return true;
+            }
+            foreach (var providerReference in this._namedProviderReferences.Values)
+            {
+                if (providerReference == provider)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private bool IsProviderActive(FeatureProviderReference providerRef)
+        {
+            return this._activeSubscriptions.Contains(providerRef);
+        }
+
+        private void EmitOnRegistration(FeatureProviderReference provider, ProviderEventTypes eventType, EventHandlerDelegate handler)
+        {
+            if (provider == null)
+            {
+                return;
+            }
+            var status = provider.Provider.GetStatus();
+
+            var message = "";
+            if (status == ProviderStatus.Ready && eventType == ProviderEventTypes.ProviderReady)
+            {
+                message = "Provider is ready";
+            }
+            else if (status == ProviderStatus.Error && eventType == ProviderEventTypes.ProviderError)
+            {
+                message = "Provider is in error state";
+            }
+            else if (status == ProviderStatus.Stale && eventType == ProviderEventTypes.ProviderStale)
+            {
+                message = "Provider is in stale state";
+            }
+
+            if (message != "")
+            {
+                handler(new ProviderEventPayload
+                {
+                    ProviderName = provider.Provider.GetMetadata().Name,
+                    Type = eventType,
+                    Message = message
+                });
+            }
+        }
+
+        private async Task ProcessFeatureProviderEventsAsync(FeatureProviderReference providerRef)
+        {
+            while (true)
+            {
+                var item = await providerRef.Provider.GetEventChannel().Reader.ReadAsync().ConfigureAwait(false);
+
+                switch (item)
+                {
+                    case ProviderEventPayload eventPayload:
+                        await this.EventChannel.Writer.WriteAsync(new Event { Provider = providerRef, EventPayload = eventPayload }).ConfigureAwait(false);
+                        break;
+                    case ShutdownSignal _:
+                        providerRef.ShutdownSemaphore.Release();
+                        return;
+                }
+            }
+        }
+
+        // Method to process events
+        private async Task ProcessEventAsync()
```

Comment:
personally I think this should be run in a dedicated thread instead of using async 🤔